distributed/worker.py

```diff
@@ -2137,24 +2141,21 @@ async def _maybe_deserialize_task(
         return function, args, kwargs
 
     @fail_hard
-    async def execute(self, key: str, *, stimulus_id: str) -> StateMachineEvent | None:
+    async def execute(self, key: str, *, stimulus_id: str) -> StateMachineEvent:
         """Execute a task. Implements BaseWorker abstract method.
 
         See also
         --------
         distributed.worker_state_machine.BaseWorker.execute
         """
-        if self.status in {Status.closing, Status.closed, Status.closing_gracefully}:
-            return None
-        ts = self.state.tasks.get(key)
-        if not ts:
-            return None
-        if ts.state == "cancelled":
-            logger.debug(
-                "Trying to execute task %s which is not in executing state anymore",
-                ts,
-            )
-            return AlreadyCancelledEvent(key=ts.key, stimulus_id=stimulus_id)
+        if self.status not in WORKER_ANY_RUNNING:
+            # This is just for internal coherence of the WorkerState; the reschedule
+            # message should not ever reach the Scheduler.
+            # It is still OK if it does though.
+            return RescheduleEvent(key=key, stimulus_id=f"worker-closing-{time()}")
```

Comment:
The two new tests remain green if you revert the first line of this block:
```python
if self.status in {Status.closing, Status.closed, Status.closing_gracefully}:
    return RescheduleEvent(key=key, stimulus_id=f"worker-closing-{time()}")
```
In this case, we have a reschedule event actually reach the scheduler.

Reply:
I'm concerned that this causes the scheduler to reschedule a task twice. First, because the worker left, and second, because the reschedule event arrives.
Why is the previous version of returning `None` not sufficient?

Edit: https://github.com/dask/distributed/blob/75ef3c9fa34e0dcef9cacff09b55a45efc576dc7/distributed/scheduler.py#L6740-L6741 scheduler verifies the reschedule event +1